lib/state/market-creation/types/validation.ts

```diff
@@ -0,0 +1,262 @@
+import { ZeitgeistPrimitivesMarketMarketCreation } from "@polkadot/types/lookup";
+import { encodeAddress } from "@polkadot/util-crypto";
+import { tryCatch } from "@zeitgeistpm/utility/dist/option";
+import { ChainTime } from "@zeitgeistpm/utility/dist/time";
+import { defaultTags } from "lib/constants/markets";
+import {
+  MarketDeadlineConstants,
+  useMarketDeadlineConstants,
+} from "lib/hooks/queries/useMarketDeadlineConstants";
+import { useChainTime } from "lib/state/chaintime";
+import { useMemo } from "react";
+import * as z from "zod";
+import { SupportedCurrencyTag } from "../../../constants/supported-currencies";
+import { minBaseLiquidity } from "../constants/currency";
+import { MarketFormData } from "./form";
+import { timelineAsBlocks } from "./timeline";
+
+export type MarketValidationDependencies = {
+  form: Partial<MarketFormData>;
+  deadlineConstants: MarketDeadlineConstants;
+  chainTime: ChainTime;
+};
+
+/**
+ * Creates a zod schema validator for the market creation form.
+ * Needs some context like the other form fields and some remote data like on
+ * chain constants and chain time.
+ *
+ * @param form - used for context where fields need to be validated against each other.
+ * @param deadlineConstants - used to validate deadlines against chain min and max constants.
+ * @param chainTime - used to validate deadlines against chain time.
+ */
+export const createMarketFormValidator = ({
+  form,
+  deadlineConstants,
+  chainTime,
+}: MarketValidationDependencies) => {
+  const timeline = timelineAsBlocks(form, chainTime).unwrap();
+
+  return z
+    .object({
+      currency: IOCurrency,
+      question: IOQuestion,
+      tags: IOTags,
+      answers: IOAnswers,
+      endDate: IOEndDate,
+      gracePeriod: IOPeriodOption.refine(() => !(timeline?.grace.period < 0), {
+        message: "Grace period must be after market end date",
+      }).refine(
+        () => !(timeline?.grace.period > deadlineConstants?.maxGracePeriod),
+        {
+          message: `Grace period must be less than ${deadlineConstants?.maxGracePeriod} blocks.`,
```

Comment:
Would be nice to covert these things to a time rather than blocks as the user is entering a time